WordPress/gutenberg
# [Draft] Use template parts to preserve navigation between theme switches

This PR preserves the navigation menu between theme switches as described in https://github.com/WordPress/gutenberg/issues/36087.

Here's a demo:

https://user-images.githubusercontent.com/205419/139687448-9a6a9b8d-1b3f-467c-aadb-9249728391df.mp4

Upsides:

This PR reuses the template parts system so users don't need to learn new mental models. The PR doesn't change the user interface in any way. Anything supported by template parts just works, e.g. the isolated template part editor. Also, the change is relatively succinct and self-contained. It introduces the concept of "applying attributes on theme switch," but not much besides that.

Downsides:
* Each navigation template part becomes a database entry on theme switch. This can be mitigated by removing unchanged template parts on theme switch.
* The navigation template parts are not technically restricted just to the navigation blocks. This can be mitigated by introducing more specialized “Slots” restricted to specific blocks.
* There is a naming convention for the navigation-related template parts (`<name>-menu.html`). This can be mitigated by listing these template parts explicitly in theme.json.

Missing:
* Support for classic theme -> Block theme transition and vice versa (easy to add)
* Guards against corner cases like an empty template part file and such


**Test plan:**

1. Have two block themes. I installed `Mayland (blocks)` and cloned it as `Mayland (blocks) 2`
2. Update them to include a `primary-menu` template part in header.html (e.g. ``) 
3. Add a default implementation of that template part by creating a `primary-menu.html ` file in `block-template-parts` of each theme. Add a navigation block to each: ``
4. Add this template part to theme.json of both themes, use `"area": "primary-menu"`
5. Go through the steps from the video above, confirm it all worked as expected.
